### PR TITLE
[FIXED JENKINS-36432] Added a SecretBytes class to act as a byte[] holder

### PR DESCRIPTION
**Pull request by stephenc**
[JENKINS-36432](https://issues.jenkins-ci.org/browse/JENKINS-36432)

@reviewbybees esp @jglick

### CODE REVIEW

**Comment by jglick** [comment]
```
java.lang.NullPointerException
	at com.cloudbees.plugins.credentials.CredentialsProviderTest.testManageUserCredentials(CredentialsProviderTest.java:154)
```

File: src/main/java/com/cloudbees/plugins/credentials/SecretBytes.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * An analogue of {@link Secret} to be used for efficient storage of {@link byte[]}. The serialized form will embed the
+ * salt and padding so no two invocations of {@link #getEncryptedData()} will return the same result, but all will
+ * decrypt to the same {@link #getPlainData()}. XStream serialization and Stapler form-binding will assume that
+ * the {@link #toString()} representation is used (i.e. the Base64 encoded secret bytes wrapped with <code>{</code>
+ * and <code>}</code>. If the string representation fails to decrypt (and is not wrapped
```
Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/SecretBytes.java]: …?

**Comment by jglick** [comment]
What is the motivation of all this? To save file size for certificate credentials? Is the added complexity worth optimizing for the relatively uncommon use case?

**Comment by stephenc** [comment]
Also to allow plain credentials to stop hand-rolling in https://github.com/jenkinsci/plain-credentials-plugin/blob/master/src/main/java/org/jenkinsci/plugins/plaincredentials/FileCredentials.java

**Comment by jglick** [comment]
So should there be a matching downstream PR in `plain-credentials-plugin` demonstrating that use case?

**Comment by stephenc** [comment]
Well there is a demo of the use case in this PR so I didn't think it was necessary

**Comment by jglick** [comment]
Well the form binding in `FileCredentialsImpl` was so hackish and fragile that I will believe changes to it work when I see them (and you have tested them in all scenarios).

**Comment by stephenc** [comment]
Hmmm I have managed to reproduce some of the test failures on linux but not of os-x

**Comment by stephenc** [comment]
```
Tests in error: 
com.cloudbees.plugins.credentials.CredentialsProviderTest.testManageUserCredentials(com.cloudbees.plugins.credentials.CredentialsProviderTest)
  Run 1: CredentialsProviderTest.testManageUserCredentials:154 » NullPointer
  Run 2: CredentialsProviderTest.testManageUserCredentials:154 » NullPointer
  Run 3: CredentialsProviderTest.testManageUserCredentials:154 » NullPointer
  Run 4: CredentialsProviderTest.testManageUserCredentials:154 » NullPointer
  Run 5: CredentialsProviderTest.testManageUserCredentials:154 » NullPointer

```

What is strange is I cannot see the NPE: https://github.com/jenkinsci/credentials-plugin/blob/jenkins-36432/src/test/java/com/cloudbees/plugins/credentials/CredentialsProviderTest.java#L154

**Comment by stephenc** [comment]
OK this is nuts... I need to investigate further... something very strange is going on here

**Comment by stephenc** [comment]
Aha!

```
java.lang.NullPointerException
        at jenkins.security.ConfidentialKey.load(ConfidentialKey.java:47)
        at jenkins.security.HMACConfidentialKey.getKey(HMACConfidentialKey.java:125)
        at jenkins.security.HMACConfidentialKey.createMac(HMACConfidentialKey.java:112)
        at jenkins.security.HMACConfidentialKey.mac(HMACConfidentialKey.java:69)
        at jenkins.security.HMACConfidentialKey.mac(HMACConfidentialKey.java:85)
        at jenkins.security.ApiTokenProperty$DescriptorImpl.newInstance(ApiTokenProperty.java:116)
        at jenkins.security.ApiTokenProperty$DescriptorImpl.newInstance(ApiTokenProperty.java:101)
        at hudson.model.User.load(User.java:192)
        at hudson.model.User.<init>(User.java:141)
        at hudson.model.User.getOrCreate(User.java:451)
        at hudson.model.User.get(User.java:402)
        at hudson.model.User.get(User.java:361)
        at hudson.model.User.get(User.java:480)
        at com.cloudbees.plugins.credentials.CredentialsProviderTest.testManageUserCredentials(CredentialsProviderTest.java:156)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.jvnet.hudson.test.JenkinsRule$2.evaluate(JenkinsRule.java:523)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```

**Comment by stephenc** [comment]
deefd5e should fix that issue

**Comment by stephenc** [comment]
let's give it a couple of rebuilds

**Comment by stephenc** [comment]
and once more

**Comment by stephenc** [comment]
Ok that fixed the test flakes

File: src/main/java/com/cloudbees/plugins/credentials/SecretBytes.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+import hudson.Util;
+import hudson.util.Secret;
+import java.io.Serializable;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import javax.crypto.Cipher;
+import jcifs.util.Base64;
```
Comment (jtnord) [inline src/main/java/com/cloudbees/plugins/credentials/SecretBytes.java]: @stephenc was this base64 encoder  intentional?

later on you use `com.trilead.ssh2.crypto.Base64` in `CertificateCredentialImpl` and even later `hudson.remoting.Base64` in the unit test.

**Pull request merged**